File: src/main/java/lec12/Lec12Main1.java
package lec12;

public class Lec12Main1 {
    public static void main(String[] args) {

        // 이렇게 companion object의 메서드 이름을 명시하여 부를 수 있고
        // 만약 이름이 따로 없다면 Person.Companion.newBaby("ABC") 이렇게 호출 가능
        Person.Factory.newBaby("ABC");

        // 만약 @JvmStatic 이라는 어노테이션을 companion object 메서드에 선언하면
        // 이렇게 바로 사용 가능
        Person.newBaby("ABC");
    }
}

File: src/main/kotlin/lec12/Lec12Main1.kt
package lec12

fun main() {

    println(Singleton.a)
    Singleton.a += 10
    println(Singleton.a)

}

class Person private constructor (
    var name: String,
    var age: Int,
) {

    // 이 companion object를 java의 static 변수와 함수 처럼 사용
    // companion object, 즉 동반객체도 하나의 객체로 간주된다. 때문에 이름을 붙일 수도 있고, interface를 구현할 수도 있다.
    companion object Factory : Log {
        private const val MIN_AGE = 1

        @JvmStatic
        fun newBaby(name: String): Person {
            return Person(name, MIN_AGE)
        }

        override fun log() {
            println("나는 Person 클래스의 동행객체 Factory에요")
        }
    }
}

object Singleton {
    var a: Int = 0
}
